File: app/src/main/java/com/kotobyte/searchpage/KanjiSearchResultsAdapter.java
package com.kotobyte.searchpage;


import android.content.Context;
import android.databinding.DataBindingUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.kotobyte.R;
import com.kotobyte.databinding.ViewKanjiItemBinding;
import com.kotobyte.models.Kanji;

import java.util.List;


class KanjiSearchResultsAdapter extends RecyclerView.Adapter<KanjiSearchResultsAdapter.ViewHolder> {

    private Context mContext;
    private Listener mListener;

    private List<Kanji> mKanjiSearchResults;

    private KanjiReadingsTextGenerator mReadingTextsGenerator;
    private KanjiMeaningsTextGenerator mMeaningTextsGenerator;

    KanjiSearchResultsAdapter(Context context, Listener listener, List<Kanji> kanjiSearchResults) {
        mContext = context;
        mListener = listener;
        mKanjiSearchResults = kanjiSearchResults;

        mReadingTextsGenerator = new KanjiReadingsTextGenerator(mContext, kanjiSearchResults);
        mMeaningTextsGenerator = new KanjiMeaningsTextGenerator(mContext, kanjiSearchResults);
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {

        ViewKanjiItemBinding binding = DataBindingUtil.inflate(LayoutInflater.from(mContext), R.layout.view_kanji_item, parent, false);
        ViewHolder viewHolder = new ViewHolder(binding);

        viewHolder.mBinding.kanjiContainer.setOnClickListener(new OnKanjiClickListener(viewHolder));

        return viewHolder;
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {

        holder.mBinding.literalTextView.setText(mKanjiSearchResults.get(position).getLiteral());
        holder.mBinding.readingsTextView.setText(mReadingTextsGenerator.getSpannableString(position));
        holder.mBinding.meaningsTextView.setText(mMeaningTextsGenerator.getSpannableString(position));
    }

    @Override
    public int getItemCount() {
        return mKanjiSearchResults != null ? mKanjiSearchResults.size() : 0;
    }

    interface Listener {
        void onClickKanji(int position, Kanji kanji);
    }

    static class ViewHolder extends RecyclerView.ViewHolder {

        private ViewKanjiItemBinding mBinding;

        ViewHolder(ViewKanjiItemBinding binding) {
            super(binding.getRoot());

            mBinding = binding;
        }
    }

    private class OnKanjiClickListener implements View.OnClickListener {

        private ViewHolder mViewHolder;

        private OnKanjiClickListener(ViewHolder viewHolder) {
            mViewHolder = viewHolder;
        }

        @Override
        public void onClick(View v) {
            int position = mViewHolder.getAdapterPosition();

            mListener.onClickKanji(position, mKanjiSearchResults.get(position));
        }
    }
}

File: app/src/main/java/com/kotobyte/search/KanjiSearchResultsAdapter.kt
package com.kotobyte.search

import android.content.Context
import android.databinding.DataBindingUtil
import android.support.v7.widget.RecyclerView
import android.text.SpannableString
import android.util.SparseArray
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.kotobyte.R
import com.kotobyte.databinding.ViewKanjiItemBinding
import com.kotobyte.models.Kanji


internal class KanjiSearchResultsAdapter(
        private val context: Context,
        private val listener: Listener,
        private val kanjiList: List<Kanji>

) : RecyclerView.Adapter<KanjiSearchResultsAdapter.ViewHolder>() {

    private val readingsTexts: List<SpannableString>
    private val meaningsTexts: List<SpannableString>

    init {
        val readingsTextGenerator = KanjiReadingsTextGenerator(context)
        val meaningsTextGenerator = KanjiMeaningsTextGenerator(context)

        readingsTexts = kanjiList.map { readingsTextGenerator.createFrom(it) }
        meaningsTexts = kanjiList.map { meaningsTextGenerator.createFrom(it) }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {

        val binding = DataBindingUtil.inflate<ViewKanjiItemBinding>(LayoutInflater.from(context), R.layout.view_kanji_item, parent, false)

        return ViewHolder(binding).apply {
            binding.kanjiContainer.setOnClickListener(OnKanjiClickListener(this))
        }
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {

        holder.binding.literalTextView.text = kanjiList[position].character.toString()
        holder.binding.readingsTextView.text = readingsTexts[position]
        holder.binding.meaningsTextView.text = meaningsTexts[position]
    }

    override fun getItemCount(): Int = kanjiList.size

    internal interface Listener {
        fun onClickKanji(position: Int, kanji: Kanji)
    }

    internal class ViewHolder(val binding: ViewKanjiItemBinding) : RecyclerView.ViewHolder(binding.root)

    private inner class OnKanjiClickListener(private val viewHolder: ViewHolder) : View.OnClickListener {

        override fun onClick(v: View) {
            val position = viewHolder.adapterPosition

            listener.onClickKanji(position, kanjiList[position])
        }
    }
}
